feat: Add bisecting tooling for nightly test regressions (#1223)
NVIDIA-NeMo/RL: docs/testing.md

Reviewer: ```suggestion
> Some unit tests require setting up test assets which you can download with: 
```

Reviewer: ```suggestion
### Refreshing Remote-Selection Artifacts
```

Author: @jgerh for these my preference is to keep the github admonition b/c we have logic inside conf.py that will translate it to the myst style. so this way the github admonition is formatted nicely as well as the official docs

> CONTAINER=... bash run_functional_in_docker.sh functional/sft.sh

Reviewer: Please verify. Earlier in the guide (under "Run Unit Tests in a Hermetic Environment"), the script is referenced as tests/run_unit_in_docker.sh.

```suggestion
CONTAINER=... bash tests/run_functional_in_docker.sh functional/sft.sh
```

### Bisecting Unit/Functional Tests

Use `tools.bisect/bisect-run.sh` to automatically run your test command across a commit range and find the first bad commit. It forces venv rebuilds so dependencies match each commit.

Basic usage:

```sh
GOOD=<good_ref> BAD=<bad_ref> \
  tools.bisect/bisect-run.sh uv run --group test pytest tests/unit/test_foobar.py::test_case
```

Examples:

```sh
GOOD=56a6225 BAD=32faafa \
  tools.bisect/bisect-run.sh uv run --group dev pre-commit run --all-files

GOOD=464ed38 BAD=c843f1b \
  tools.bisect/bisect-run.sh uv run --group test pytest tests/unit/test_foobar.py
```

Notes:

- Exit codes drive the classification: 0=good, non-zero=bad, 125=skip.
- The script pre-verifies that `GOOD` is actually good by running your command on it.
- On failure or interruption, it saves a timestamped `git bisect log` to `<repo>/bisect-logs/`. You can resume later with `BISECT_REPLAY_LOG` (see below).
- Set `BISECT_NO_RESET=1` to keep the bisect state after the script exits.

Resume from a saved bisect log:

```sh
BISECT_REPLAY_LOG=/abs/path/to/bisect-2025....log \
  tools.bisect/bisect-run.sh uv run --group test pytest tests/unit/test_foobar.py
```

> ### Bisecting nightlies

Reviewer: ```suggestion
### Bisecting Nightlies
```

Nightly training scripts can be bisected using the same driver plus a helper that sets up hermetic runs on Slurm.

Vanilla flow:

```sh
# Copy bisect utilities outside of VCS to ensure a stable runner
rsync -ahP --delete tools/ tools.bisect/

TEST_CASE=tests/test_suites/llm/sft-llama3.2-1b-1n8g-fsdp2tp1.v3.sh

HF_HOME=... \
HF_DATASETS_CACHE=... \
CONTAINER=... \
MOUNTS=... \
ACCOUNT=... \
PARTITION=... \
GOOD=$(git log --format="%h" --diff-filter=A -- "$TEST_CASE") \
BAD=HEAD \
  tools.bisect/bisect-run.sh tools.bisect/launch-bisect.sh "$TEST_CASE"
```

::::{note}
The command `GOOD=$(git log --format="%h" --diff-filter=A -- "$TEST_CASE")` selects the commit that introduced the test script. Because the path is typically added only once, this yields the introduction commit to use as the known good baseline.
::::

- `tools.bisect/launch-bisect-helper.sh` ensures each commit runs in a fresh venv, creates an isolated code snapshot per commit, blocks until metrics are checked, and returns a suitable exit code for bisect.

Progressively more advanced cases:

1) Adjusting the test case on the fly with `SED_CLAUSES`

> - If a test script needs small textual edits during bisect (e.g., relax a threshold; drop a noisy metric you don’t care to bisect over when focusing on convergence vs. perf), provide a sed script via `SED_CLAUSES`. You can also use this to adjust runtime controls like `MAX_STEPS`, `STEPS_PER_RUN`, or `NUM_MINUTES` when a perf regression slows runs down so they still complete and emit metrics. The helper applies it and automatically restores the test script after the run.

Reviewer: ```suggestion
- If a test script needs small textual edits during bisect (e.g., to relax a threshold or drop a noisy metric you don't care to bisect over when focusing on convergence vs. performance), provide a sed script via `SED_CLAUSES`. You can also use this to adjust runtime controls like `MAX_STEPS`, `STEPS_PER_RUN`, or `NUM_MINUTES` when a performance regression slows runs down, ensuring they still complete and emit metrics. The helper applies it and automatically restores the test script after the run.
```
